packages/teleport/src/lib/tdp/client.ts

```diff
@@ -0,0 +1,148 @@
+// Copyright 2021 Gravitational, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import { EventEmitter } from 'events';
+import Codec, { MessageType, MouseButton, ButtonState } from './codec';
+import Logger from 'shared/libs/logger';
+
+// Client is the TDP client. It is responsible for connecting to a websocket serving the tdp server,
+// sending client commands, and recieving and processing server messages.
+export default class Client extends EventEmitter {
+  codec: Codec;
+  socket: WebSocket;
+  socketAddr: string;
+  username: string;
+  logger = Logger.create('TDPClient');
+  userDisconnected = false;
+
+  constructor(socketAddr: string, username: string) {
+    super();
+    this.socketAddr = socketAddr;
+    this.codec = new Codec();
+    this.username = username;
+  }
+
+  // Connect to the websocket and register websocket event handlers.
+  init() {
+    this.socket = new WebSocket(this.socketAddr);
+    this.socket.binaryType = 'arraybuffer';
+
+    this.socket.onopen = () => {
+      this.logger.info('websocket is open');
+      this.emit('init');
+    };
+
+    this.socket.onmessage = (ev: MessageEvent) => {
+      this.processMessage(ev.data as ArrayBuffer);
+    };
+
+    // The 'error' event will only ever be emitted by the socket
+    // prior to a 'close' event (https://stackoverflow.com/a/40084550/6277051).
+    // Therefore, we can rely on our onclose handler to account for any websocket errors.
+    this.socket.onerror = null;
+    this.socket.onclose = () => {
+      this.logger.info('websocket is closed');
+
+      // Clean up all of our socket's listeners and the socket itself.
+      this.socket.onopen = null;
+      this.socket.onmessage = null;
+      this.socket.onclose = null;
+      this.socket = null;
+
+      if (this.userDisconnected) {
```

Comment:
just unsubscribe from 'disconnect` event from the caller. Also, by the time this event gets fired, the component would be gone (and automatically unsubscribed from all events if done correctly). Yeah, this flag seems unnessary.

Reply:
Do you mean I should unsubscribe from the `error` event in the caller ([emitted by `socket.onclose`](https://github.com/gravitational/webapps/blob/desktop-access/packages/teleport/src/lib/tdp/client.ts#L65)), before calling `client.disconnect()`? I think that's a worse API, because the caller needs to understand more about the internal implementation of the `client` to use it. With this design, you can just call `disconnect()` from anywhere and expect the `disconnect` event listener to run and no others.

Also, I don't think that by the time the `error` event gets fired, the component would be gone and the `error` event unsubscribed, since as currently written the `TdpClientCanvas` [has it's display set to `none`](https://github.com/gravitational/webapps/blob/desktop-access/packages/teleport/src/DesktopSession/DesktopSession.tsx#L112-L115), but that doesn't trigger the `useEffect` hook's cleanup function (where events get unsubscribed).

Perhaps I'm misunderstanding, but from my current perspective it seems like I'd need to do a lot of refactoring for a worse api.